src/validation.cpp

```diff
@@ -417,21 +418,135 @@ static bool CheckInputsFromMempoolAndCache(const CTransaction& tx, CValidationSt
+            std::vector<COutPoint>& coins_to_uncache, bool test_accept);
+
+private:
+    // We put the arguments we're handed into a struct, so we can pass them
+    // around easier.
+    struct ATMPArgs {
+        const CChainParams& chainparams;
+        CValidationState &state;
+        bool* pfMissingInputs;
+        int64_t nAcceptTime;
+        std::list<CTransactionRef>* plTxnReplaced;
+        bool bypass_limits;
+        const CAmount& nAbsurdFee;
+        std::vector<COutPoint>& coins_to_uncache;
+        bool test_accept;
+    };
+
+    // All the intermediate state that gets passed between the various levels
+    // of checking a given transaction.
+    struct Workspace {
+        Workspace(const CTransactionRef& ptx) : hash(ptx->GetHash()) {}
+        std::set<uint256> setConflicts;
+        CTxMemPool::setEntries allConflicting;
+        CTxMemPool::setEntries setAncestors;
+        std::unique_ptr<CTxMemPoolEntry> entry;
+
+        bool fReplacementTransaction{false};
+        CAmount nModifiedFees{0};
+        CAmount nConflictingFees{0};
+        size_t nConflictingSize{0};
+
+        const uint256& hash;
+    };
+
+    // Run the policy checks on a given transaction, excluding any script checks.
+    // Looks up inputs, calculates feerate, considers replacement, evaluates
+    // package limits, etc. As this function can be invoked for "free" by a peer,
+    // only tests that are fast should be done here (to avoid CPU DoS).
+    bool PreChecks(ATMPArgs& args, const CTransactionRef& ptx, Workspace& ws) EXCLUSIVE_LOCKS_REQUIRED(cs_main, pool.cs);
+
+    // Run the script checks using our policy flags. As this can be slow, we should
+    // only invoke this on transactions that have otherwise passed policy checks.
+    bool PolicyScriptChecks(ATMPArgs& args, const CTransactionRef& ptx, Workspace& ws, PrecomputedTransactionData& txdata) EXCLUSIVE_LOCKS_REQUIRED(cs_main);
+
+    // Re-run the script checks, using consensus flags, and try to cache the
+    // result in the scriptcache. This should be done after
+    // PolicyScriptChecks(). This requires that all inputs either be in our
+    // utxo set or in the mempool.
+    bool ConsensusScriptChecks(ATMPArgs& args, const CTransactionRef& ptx, Workspace& ws, PrecomputedTransactionData &txdata) EXCLUSIVE_LOCKS_REQUIRED(cs_main);
+
+    // Try to add the transaction to the mempool, removing any conflicts first.
+    // Returns true if the transaction is in the mempool after any size
+    // limiting is performed, false otherwise.
+    bool Finalize(ATMPArgs& args, const CTransactionRef& ptx, Workspace& ws) EXCLUSIVE_LOCKS_REQUIRED(cs_main, pool.cs);
+
+    // Compare a package's feerate against minimum allowed.
+    bool CheckFeeRate(size_t package_size, CAmount package_fee, CValidationState& state)
+    {
+        CAmount mempoolRejectFee = pool.GetMinFee(gArgs.GetArg("-maxmempool", DEFAULT_MAX_MEMPOOL_SIZE) * 1000000).GetFee(package_size);
```

Comment:
Why are some of the settings parsed in the constructor and saved as a member and others are not?

Could do the same here.

```suggestion
        CAmount mempoolRejectFee = pool.GetMinFee(m_max_pool_size).GetFee(package_size);
```